---
title: "BMIN503/EPID600 Project"
author: "Alexa Woodward"
output: 
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 400)
```  
***

## Overview
The goal of this project is to build a machine learning model that can predict gestational age based on differentially methylated regions in preterm and term newborns. Publicly available methylation array data from the Gene Expression Omnibus (GEO) will be used for this purpose. Quality control will be performed using standard approaches prior to feature selection, and subsequently a support vector machine (SVM) will be used to generate the model. We hypothesize that differential methylation will be a good predictor of gestational age, which has implications for research considering making causal associations using the methylome of these two groups.

# https://github.com/alexa-woodward/BMIN503_Final_Project.git 


## Introduction 
Epigenetics is a burgeoning field that is making a concerted effort to explain the mechanisms behind gene-environment interactions, the developmental origins of health and disease, cancer, and other phenomena [1-4]. Perinatal stressors have emerged as key exposures that impact later life outcomes, especially neurodevelopment, metabolic diseases, and obesity [5-7]. Conrad Waddington, a embryologist and developmental biologist, is creditied with coining the term "epigenetics" in 1942 [8]. The exact defition is contested, but generally refers to the elements that modify the expression of genes without changing the DNA sequence itself. As the genetic era began soon thereafter, new questions about differences in gene expression and phenotype development, the discovery of imprinted genes, and developments in sequencing and analysis methods paved the way for epigenetics research. More recently, epigenetic marks have been shown to be influenced by a number of perinatal exposures, including maternal behavior [3], famine [9], time spent in the neo-natal intensive care unit (NICU) [10], and preterm birth [11].

Unlike classical genetics and despite the wealth of studies in animal models and continuing work in birth cohorts and other human samples, epigenetic research has not yet made use of machine learning. In genetics, machine learning has been applied extensively as a tool for prediction, dimensionality reduction, and classification. It has provided information on population and individual risk, and has increased the accuracy of various predictions [12-14]. Implementing these methods using epigenetic research is a novel approach will require interdisciplinary input from data scientists, basic science researchers in genetics and epigenetics, clincians, evironmental health scientists, epidemiologists and others. This study will be the first of its kind to use machine learning in this context of prenatal exposures and epigenetics in a human sample.

The methylome is highly variable during early development [15, 16], and thus gestational age may confound relationships between differential methylation and observed outcomes. Even aside from any apparent pathology, gestational age may explain a large portion of the differential methylation in these subjects.

## Study and Control Population
Study population data were downloaded from GEO, experiment GSE72120. The data is comprised of methylation arrays from saliva samples n from 36 preterm and 36 term infants who recieved care at the Royal Infirmary of Edinburgh between 2012 and 2014. Preterm birth was defined as < 32 weeks gestation, and term birth as > 37 weeks gestation. In each group, there are 18 males and 18 females. For the preterm infants, data on steriod administration (yes/no) were also included. fMRI variables will not be included in this analysis.


---
title: "Woodward_FinalProject"
author: "Alexa Woodward"
date: "November 5, 2018"
output: html_document
---

# Required Packages
```{r eval=FALSE}

if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()
BiocManager::install(c("GenomicFeatures", "AnnotationDbi"))
BiocManager::install("IlluminaHumanMethylation450kprobe", version = "3.8")
BiocManager::install("IlluminaHumanMethylation450kmanifest", version = "3.8")
BiocManager::install(c("IlluminaHumanMethylation27k.db"))
BiocManager::install("GEOquery", version = "3.8")
BiocManager::install("openssl")
BiocManager::install("minfi")
BiocManager::install("minfiData")
BiocManager::install("missMethyl")
BiocManager::install("bumphunter", version = "3.8")
BiocManager::install("shinyMethyl")
BiocManager::install("methylationArrayAnalysis", version = "3.8")
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
BiocManager::install("methyAnalysis")
install.packages("corrplot")
install.packages("mice")
install.packages("VIM")
install.packages("caret")
```


## Methylation Array - Illumnia 450K
DNA from the saliva samples was hybridized to the Illumina Infinium 450k Human Methylation Beadchip. This array includes two assays, the Infinium I and Infinium II, which use two and one bead(s) per probe, respectively. Methylated and unmethylated sites are measured by intensity values from "red" (A/T) and "green" (G/C) fluoresence. For each locus, the methylated proportion can be caluculated from these intensities, represented as M or Beta-values. 

### Get GEO supplementary files, read in data
```{r}
library(GEOquery)
library(minfi)
getGEOSuppFiles("GSE72120")

untar("GSE72120/GSE72120_RAW.tar", exdir = "C:/Users/alexaw/Documents/EPID 600/GSE72120/idat")
head(list.files("C:/Users/alexaw/Documents/EPID 600/GSE72120/idat", pattern = "idat"))

idatFiles <- list.files("C:/Users/alexaw/Documents/EPID 600/GSE72120/idat", pattern = "idat.gz$", full = TRUE)
sapply(idatFiles, gunzip, overwrite = TRUE)

#create target file
baseDir_array <- file("C:/Users/alexaw/Documents/EPID 600/GSE72120/idat")

targetarray <- read.metharray.sheet("C:/Users/alexaw/Documents/EPID 600/GSE72120/idat")

library(illuminaio)
rg_array <- read.metharray.exp("C:/Users/alexaw/Documents/EPID 600/GSE72120/idat", targets = targetarray)
rg_array
#create MethylSet
mset_array <- preprocessRaw(rg_array)


#probe info
probe_info <- getManifest(rg_array)
probe_info
head(getProbeInfo(probe_info))

#annotation
annotation <- getAnnotation(rg_array)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
annot450K <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
                           
```

## Methods
Using methylation array data from GEO, this pipeline will perform QC, allow for visualization, and ultimately use machine learning methods to show if differentially methylated sites can predict gestational age.

## Data preprocessing and quality control
Data downloaded using the GEOquery package from Bioconductor. IDAT files were downloaded and gzipped to a local directory. The matrix file was modified to produce a target file with relevant phenotype information. Probe information was obtained and the array was annotated using the minfi package and dependencies. 
minfi [18]
 
### Normalization: The data was normalized using the preprocessFunnorm, recommended for large-scale baseline differences between samples [17]. Probes spanning SNPs were removed from the data.
```{r}
library(minfi)
library(shinyMethyl)
summary_array <- shinySummarize(rg_array)

##phenotype data
pheno <- pData(rg_array)

#getQC
qc_info <- getQC(mset_array)
plotQC(qc_info, badSampleCutoff = 10)

#normalize 
array_norm <- preprocessFunnorm(rg_array)
head(array_norm)
summary_norm <- shinySummarize(array_norm)
head(summary_norm@phenotype)
```


## Data Visualization
```{r eval=FALSE}
#Run ShinyMethyl for visualization and QC
runShinyMethyl(summary_array, summary_norm) #compares normal and non-normalized data
```

### Remove SNPs
```{r}
array_snps <- getSnpInfo(array_norm)
array_qc <- dropLociWithSnps(array_norm, snps=c("SBE","CpG"), maf=0)

```


## Differential Methylation: The minfi and bumphunter packages were utilized to find differentially methylated positions and regions (DMRs) in the dataset."Positions" refer to individual CpG probes, while "regions" over a number of bases can have one or more probes. In finding DMRs, the data was resampled 100 times, then the output was annotated using the list of known genes in hg19 from the University of California, Santa Cruz. A total of 274 DMRs were identified. 

```{r}
#dmpFinder for differentially methylated positions
beta_array <- getBeta(array_qc) #no warning here
casecontrol <- pData(array_qc)$casecontrolstatus
dmp_casecont <- dmpFinder(beta_array, pheno = casecontrol, type = "categorical")
head(dmp_casecont)


#bumphunter for differentially methylated regions 
matrix <- model.matrix(~ casecontrol)
dmrs <- bumphunter(array_qc, design = matrix, 
             cutoff = 0.1, B=0, type="Beta") #can mess with these cutoff values

#run permutations
dmrs <- bumphunter(array_qc, design = matrix, 
             cutoff = 0.10, B=100, type="Beta")
dmrs2 <- as.data.frame(dmrs$table)
#annotate DMRs
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
hg19_annotated <- annotateTranscripts(TxDb.Hsapiens.UCSC.hg19.knownGene, by = c("gene"))
dmrs_anno <- annotateNearest(dmrs$table, hg19_annotated)
head(dmrs_anno)

annoTx <- annotateTranscripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
match <- matchGenes(dmrs$table, annoTx)

#Visualizing a specific DMP 
plotCpg(mset_array, cpg = "cg13985639", pheno = targetarray$casecontrolstatus, type = c("categorical"),
  measure = c("beta"))
```


### Pull potential features from phenotype data 
```{r}
#make dataframe with potential features
featurelist <-  as.data.frame(targetarray[2:9])
featurelist[,2:4] <- NULL 
featurelist$SampleID <- as.factor(featurelist$SampleID)
featurelist$Sex <- as.factor(featurelist$Sex)
featurelist$pma_birth <- as.factor(featurelist$pma_birth)
featurelist$Steriods_birth <- as.factor(featurelist$Steriods_birth)

#get M (intensity) values
Mval_cpg <- as.data.frame(getM(array_qc))
```


### Imputation of Missing Data Values: Post-menstrual age (pma) at birth, the variable of interest (a measure of gestational age), was missing for a number of the controls. However, due to the relative homogenity of values among term infants (defined as > 38 weeks) we proceeded with the assumption that this data was MCAR, and performed imputation for these values using the MICE package.
```{r}
# impute the missing PMA_birth data using MICE
library(mice)
md.pattern(featurelist)

library(VIM)
aggr_plot <- aggr(featurelist, col=c('navyblue','red'), numbers=TRUE, sortVars=TRUE, labels=names(featurelist), cex.axis=.7, gap=3, ylab=c("Histogram of missing data","Pattern"))

temp <- mice(featurelist, m=5,maxit=50,meth='pmm',seed=500)
summary(temp)
temp$imp$pma_birth

imputed <- complete(temp,1)
imputed$casecontrolstatus <- as.factor(imputed$casecontrolstatus)
```

### Create PMA "levels" for prediction using a SVM
```{r}
#create new variable for levels of Pma @birth 
imputed$pma_levels <- as.numeric(as.character(imputed$pma_birth))
imputed$pma_levels[imputed$pma_levels > 23.00 & imputed$pma_levels < 26] <- 1
imputed$pma_levels[imputed$pma_levels >= 26.00 & imputed$pma_levels < 29] <- 2
imputed$pma_levels[imputed$pma_levels > 29.00 & imputed$pma_levels < 33] <- 3
imputed$pma_levels[imputed$pma_levels > 38.00] <- 4
imputed$pma_levels <- as.factor(imputed$pma_levels)

#NOTE: The defintion for preterm for this study was defined as <32 weeks gestation, but two infants were slighly above 32 weeks (32, and 32.71 respectively) which was two few to use for prediction. Thus, level "3" was extended to 33 to allow for these subjects to stay in the dataset. 

```


### Filter cpgs from differentially methylated regions
```{r}
library(dplyr)
cpgs <- as.data.frame(array_qc@rowRanges@ranges)

cpgs_start <- inner_join(cpgs, dmrs2, by = "start")

#filter M_beta for those cpgs
library(tibble)
Mval_cpg <- rownames_to_column(Mval_cpg, var = "names")
cpg_features <- cpgs_start[,c("names", "value", "area", "p.value")]
dmr_cpgs <- inner_join(Mval_cpg, cpg_features, by = "names")

#transpose & fix column/row names
dmr_cpgs <- t(dmr_cpgs)
colnames(dmr_cpgs) <- unlist(dmr_cpgs[row.names(dmr_cpgs)=='names',])
dmr_cpgs <- dmr_cpgs[!row.names(dmr_cpgs)=='names',]
dmr_cpgs <- as.data.frame(dmr_cpgs)
dmr_cpgs <- rownames_to_column(dmr_cpgs, var = "SampleID")
dmr_cpgs$SampleID <- gsub("_.*", "",dmr_cpgs$SampleID)
```


### Join CpG M values with Phenotype data
```{r}
#add as features 
features2 <- inner_join(dmr_cpgs, imputed, by = "SampleID")
features2[,276:279] <- NULL #keep only pma_birth
features2 <- column_to_rownames(features2, var = "SampleID")
features2[,1:274] <- sapply(features2[,1:274], as.character)
features2[,1:274] <- sapply(features2[,1:274], as.numeric)

#check for any missingness
is.na(features2)

```

## Results
Below details code for the support vector machine used to predict gestational age categories using only the 274 DMRs identified above. Prior to cross-validation, the model was perfectly predictive, and was over-fit. 5-Fold cross-validation produced a model wtih 78% accuracy, and a kappa statistic of 0.65, suggestive of "substantial agreement" [19]. Additionally, the positive predictive value(PPV) and negative predicitve value (NPV) were highest for groups 1 and 4 (1, 0.96; 1, 1). PPV and NPV and were lowest among group 3 (0.46, 0.9), but this group also had the second lowest number of samples. The full output for reliability if produced by the confusion matrix.

## Support Vector Machine
```{r}
library(e1071)
GSE72120.svm <- svm(pma_levels ~ ., data=features2, scale=TRUE, kernel="linear")
GSE72120.svm

svm.pred <- fitted(GSE72120.svm)
table(features2$pma_levels, svm.pred)

svm.pred <- predict(GSE72120.svm, features2)
```


### K-Fold Cross Validation
```{r}
N = nrow(features2)
K = 5
set.seed(1234)
s = sample(1:K, size=N, replace=T)
pred.outputs.svm <- vector(mode="numeric", length=N)
obs.outputs <- vector(mode="numeric", length=N)
offset <- 0
for(i in 1:K){
    train <- filter(features2, s != i)
    test <- filter(features2, s == i)
    obs.outputs[1:length(s[s==i]) + offset] <- test$pma_levels

    #SVM train/test
    svm.m <- svm(pma_levels~., data=train, scale=TRUE, kernel="linear")
    svm.pred.curr <- predict(svm.m, test) 
    pred.outputs.svm[1:length(s[s==i]) + offset] <- svm.pred.curr
    
    offset <- offset + length(s[s==i])
}
    
```

### Check reliability
```{r}
library(caret)
obs.outputs2 <- as.factor(obs.outputs)
pred.outputs.svm2 <- as.factor(pred.outputs.svm)
confusionMatrix(pred.outputs.svm2, obs.outputs2)
```
## Secondary analyses: The convincing results from the SVM prompted a further look at the features used for prediction. A list of 20 CpGs was generated by filtering the most highly significant CpGs from the original list of 274. These 20 were used as components for PCA. WHile this analysis is incomplete, the output shows that the first 5 components explain nearly 70% of the observed variation.

### Prinicpal Components Analysis
```{r}
library(dplyr)
library(ggplot2)
library(ggbiplot)

sigcpgs <- filter(cpg_features, cpg_features$p.value < 0.05)
listsigcpgs <- list(sigcpgs$names)

features3 <- select(features2, "cg27139851","cg25456593", "cg24067911", "cg23956190", "cg21945949", "cg20695611", "cg20017683", "cg18232235", "cg16596052", "cg16444641", "cg15415945", "cg09556042", "cg08567941", "cg08349400", "cg07651316", "cg04118102", "cg02084912", "cg01752594", "cg00529567", "cg00288598", "pma_levels")
str(features3)


features.pca <- prcomp(features3[, 1:20], scale=TRUE)
print(features.pca)
summary(features.pca)

ggbiplot(features.pca, groups=features2$pma_levels, circle=TRUE, ellipse=TRUE) +
    scale_color_discrete(name = '') +
    theme(legend.direction = 'horizontal', legend.position = 'top')
```


## Limitations: First, a number of the PMA values were assumed to be MCAR and were imputed. This may have impacted the model and over-emphazised the perfect prediction of gestational age for infants > 38 weeks. Not all of the DMRs included int the SVM had a statistically significant p-value, and this could be addressed by re-running the SVM using only the CpGs that were included in the PCA. Lastly, this methylation data was initally normalized using a function specifically to account for baseline differences between groups. If the more traditional quantile normalization had been used, the prediction model may be been more accurate.

## Conclusions
DMRs are strongly predictive of gestational age using a SVM. This may suggest that these DMRs found in comparisons of term and preterm infants may be largely the result of different stages of development, rather than of any potential pathology. This model and hypothesis will need to be validated using additional datasets as well as other machine learning algorithms (e.g., random forests).


## References
1. Godfrey, Keith M., Karen A. Lillycrop, Graham C. Burdge, Peter D. Gluckman, and Mark A. Hanson. "Epigenetic mechanisms and the mismatch concept of the developmental origins of health and disease." Pediatric research 61, no. 5 Part 2 (2007): 5R.
2. Yamada, L., and S. Chong. "Epigenetic studies in developmental origins of health and disease: pitfalls and key considerations for study design and interpretation." Journal of developmental origins of health and disease 8, no. 1 (2017): 30-43.
3. Weaver, Ian CG, Nadia Cervoni, Frances A. Champagne, Ana C. D'Alessio, Shakti Sharma, Jonathan R. Seckl, Sergiy Dymov, Moshe Szyf, and Michael J. Meaney. "Epigenetic programming by maternal behavior." Nature neuroscience 7, no. 8 (2004): 847.
4. Bind, Marie-Abele, Johanna Lepeule, Antonella Zanobetti, Antonio Gasparrini, Andrea A. Baccarelli, Brent A. Coull, Letizia Tarantini, Pantel S. Vokonas, Petros Koutrakis, and Joel Schwartz. "Air pollution and gene-specific methylation in the Normative Aging Study: association, effect modification, and mediation analysis." Epigenetics 9, no. 3 (2014): 448-458.
5. Kundakovic, Marija, and Ivana Jaric. "The epigenetic link between prenatal adverse environments and neurodevelopmental disorders." Genes 8, no. 3 (2017): 104.
6. O’donnell, K., T. G. O’connor, and V. Glover. "Prenatal stress and neurodevelopment of the child: focus on the HPA axis and role of the placenta." Developmental neuroscience 31, no. 4 (2009): 285-292.
7. Welberg, Leonie AM, and Jonathan R. Seckl. "Prenatal stress, glucocorticoids and the programming of the brain." Journal of neuroendocrinology 13, no. 2 (2001): 113-128.
8. Goldberg, Aaron D., C. David Allis, and Emily Bernstein. "Epigenetics: a landscape takes shape." Cell 128, no. 4 (2007): 635-638.
9. Heijmans, Bastiaan T., Elmar W. Tobi, Aryeh D. Stein, Hein Putter, Gerard J. Blauw, Ezra S. Susser, P. Eline Slagboom, and L. H. Lumey. "Persistent epigenetic differences associated with prenatal exposure to famine in humans." Proceedings of the National Academy of Sciences 105, no. 44 (2008): 17046-17049.
10. Montirosso, Rosario, and Livio Provenzi. "Implications of epigenetics and stress regulation on research and developmental care of preterm infants." Journal of Obstetric, Gynecologic & Neonatal Nursing 44, no. 2 (2015): 174-182.
11. Burris, Heather H., Andrea A. Baccarelli, Robert O. Wright, and Rosalind J. Wright. "Epigenetics: linking social and environmental exposures to preterm birth." Pediatric research 79, no. 1-2 (2015): 136.
12. Chen, Hsinchun. "Machine learning for information retrieval: neural networks, symbolic learning, and genetic algorithms." Journal of the American society for Information Science 46, no. 3 (1995): 194-216.
13. Joubert, Bonnie R., Siri E. Håberg, Roy M. Nilsen, Xuting Wang, Stein E. Vollset, Susan K. Murphy, Zhiqing Huang et al. "450K epigenome-wide scan identifies differential DNA methylation in newborns related to maternal smoking during pregnancy." Environmental health perspectives 120, no. 10 (2012): 1425.
14. Libbrecht, Maxwell W., and William Stafford Noble. "Machine learning applications in genetics and genomics." Nature Reviews Genetics 16, no. 6 (2015): 321.
15. Geiman, Theresa M., and Kathrin Muegge. "DNA methylation in early development." Molecular Reproduction and Development: Incorporating Gamete Research 77, no. 2 (2010): 105-113.
16. Borgel, Julie, Sylvain Guibert, Yufeng Li, Hatsune Chiba, Dirk Schübeler, Hiroyuki Sasaki, Thierry Forné, and Michael Weber. "Targets and dynamics of promoter DNA methylation during early mouse development." Nature genetics 42, no. 12 (2010): 1093.
17. Jean-Philippe Fortin, Aurélie Labbe, Mathieu Lemire, Brent W Zanke, Thomas J Hudson, Elana J Fertig, Celia MT Greenwood, and Kasper D Hansen. 2014. “Functional normalization of 450k methylation array data improves replication in large cancer studies.” Genome Biology 15 (11): 503. doi:10.1186/s13059-014-0503-2.
18. Aryee MJ, Jaffe AE, Corrada-Bravo H, Ladd-Acosta C, Feinberg AP, Hansen KD, Irizarry RA (2014). “Minfi: A flexible and comprehensive Bioconductor package for the analysis of Infinium DNA Methylation microarrays.” Bioinformatics, 30(10), 1363–1369. doi: 10.1093/bioinformatics/btu049. 
19. Landis JR, Koch GG (1977) The measurement of observer agreement for categorical data. Biometrics, 33: 159-174. 